session/promise/storage.go

```diff
@@ -142,6 +147,45 @@ func (s *Storage) GetAllKnownIssuers() []identity.Identity {
 	return res
 }
 
+var errNoPromiseForConsumer = errors.New("no promise for consumer")
+
+// FindPromiseForConsumer returns the last known promise for the issuer/consumer combo
+// It checks if any promises match, and if they do checks to see if any later promises have been cleared.FindPromiseForConsumer
+func (s *Storage) FindPromiseForConsumer(issuerID, consumerID identity.Identity) (StoredPromise, error) {
+	s.Lock()
+	defer s.Unlock()
+	promises, err := s.getAllPromisesForIssuer(issuerID)
+	if err != nil {
+		return StoredPromise{}, err
+	}
+
+	// sort by sequenceID, descending
+	sort.Slice(promises, func(i, j int) bool {
+		return promises[i].SequenceID > promises[j].SequenceID
+	})
+
+	// Iterate from the last promise to the first
+	for i := 0; i < len(promises); i++ {
+		// if we find a cleared promise, it means we've done our job here - we'll need to issue a new id
+		if promises[i].Cleared {
+			return StoredPromise{}, errNoPromiseForConsumer
+		}
+		// otherwise, we're free to extend
+		if promises[i].ConsumerID == consumerID {
+			return promises[i], nil
+		}
+	}
+
+	return StoredPromise{}, errNoPromiseForConsumer
+}
+
+func (s *Storage) sortPromisesDesc(promises []StoredPromise) {
```

Comment:
Not sure if it should be a method of the `*Storage`. It feels like we are sorting it inside the storage.

Reply:
fixd